### PR TITLE
Array flatten in place

### PR DESCRIPTION
**Pull request by TheGrizzlyDev**
Partially implements Array::flatten! from #39 with the exception of:
- Make use of respond_to_missing and method_missing in value for when respond_to would otherwise return false. These methods seem to be more something we should implement in Value and in a separate PR/Issue, but if preferred I'll gladly cover both in this PR.

Followed suggestion from #75 and replaced vector with Hashmap.

### CODE REVIEW

File: src/array_value.cpp [hunk 1]
```diff
@@ -341,8 +341,18 @@ ValuePtr ArrayValue::first(Env *env, ValuePtr n) {
...
-    ArrayValue *array = new ArrayValue(klass());
+    ArrayValue *copy = new ArrayValue { *this };
+    copy->flatten_in_place(env, depth);
+    return copy;
+}
+
+ValuePtr ArrayValue::flatten_in_place(Env *env, ValuePtr depth) {
+    if (is_frozen()) {
+        env->raise("FrozenError", "can't modify frozen Array: {}", inspect_str(env));
+        return nullptr;
```
Comment (seven1m) [inline src/array_value.cpp:352]: No need to return anything here. `Env::raise()` is no return.

File: include/tm/vector.hpp [hunk 2]
```diff
@@ -100,21 +100,29 @@ class Vector {
...
+        m_data[index] = val;
     }
 
     T pop_front() {
-        assert(m_size != 0);
-        T val = m_data[0];
+        return pop_at(0);
+    }
+
+    T pop_at(size_t index) {
```
Comment (seven1m) [inline include/tm/vector.hpp:121]: I guess it's personal preference, but I've never seen a Vector `pop_at` method before and its use does seem limited. You're not using the "popped" value in your code anyway, so let's rename this to `remove()` and have it be a void return.

**Review by seven1m — CHANGES_REQUESTED** [review]
This is looking great. Just a couple small changes and then I'll merge...

File: include/tm/vector.hpp (hunk 2 above)
Comment (TheGrizzlyDev) [reply include/tm/vector.hpp:121]: I agree and it makes even more sense since I've ranamed it's counterpart to insert, thanks!

**Comment by TheGrizzlyDev** [comment]
> This is looking great. Just a couple small changes and then I'll merge...

Nice, I've addressed both comments, let me know what you think :)

**Pull request merged**